aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -217,12 +238,20 @@ void TensorIterator::compute_types() {
           op.dtype = op.tensor.scalar_type();
         } else {
           op.device = common_device;
-          op.dtype = common_dtype;
+          if (compute_common_dtype_only_for_inputs && op.is_output) {
```

Comment:
nit: this all makes sense, but this function is really hard to read due to the multiple branches :/. We should clean it up sometime.